### DISCUSSION TITLE
EntityRuler patterns not loading when using spaCy v3 config file and disabling pipe during training

### DISCUSSION THREAD

**Discussion by einarbmag** [explosion/spaCy#7684] [Help: Coding & Implementations]
I'm trying to specify EntityRuler patterns in the config file like so:
```
[initialize.components.entity_ruler.patterns]
@readers = "srsly.read_jsonl.v1"
path = "corpus/entity_ruler_patterns.jsonl"
```
but the patterns just don't seem to be loading. If I debug right after initialization, the _entity_ruler_ is in the pipeline but without any patterns. The persisted final model also has no patterns. 

Am I doing something wrong? I tried setting the patterns in the component config instead but it threw an error. Is there any other way to specify the EntityRuler patterns?

**Comment by einarbmag** [comment]
Found the issue, and I think it's a bug. I had disabled the entity_ruler pipe during training, and it seems its not initialized if it's disabled. I suppose I don't really need to disable it, but I think the behaviour is not as it should be.

**Reply by polm** [reply]
I'm not sure I understand. Do you expect the entity ruler to have the patterns in it but not execute them if it's disabled? The end result is the same.

**Comment by einarbmag** [comment]
I'm adding various components in the config file, built in and custom ones.
Some of the custom ones are perhaps not perfectly coded so they may be slow
and may throw errors so I want them disabled during training. I decided to
disable everything except tok2vec, ner and textcat and then just enable all
disabled pipes after training.

There may be other ways to achieve this. But since its possible to disable
pipes in the config file, you'd expect they're still properly configured so
they work as expected when they're enabled at runtime.

**Reply by polm** [reply]
Ah, that makes sense, though I think it's not a usage pattern we'd considered. 

There are a couple of ways to workaround it, but basically I would recommend training without the components you currently disable, and then that you add them back later.

The most straightforward way to do this right now is in code, by adding them after you load the pipeline. An alternative recently came up in [this thread](https://github.com/explosion/spaCy/discussions/7688) - you can have a separate config that loads the assembled model. This also isn't supported properly yet, but you can train for one epoch or use the existing `train` code as a base to make a custom assembler function.

I need to look at the code more, but you might be right about it being a bug, or at least undesirable behavior. I believe that not calling initialize is consistent with other components, but here it does mean that the rules just kind of vanish.

**Comment by einarbmag** [comment]
Thanks, although my hopes and dreams are to rely entirely on the config
file and spacy cli, I currently have to use a separate script to enable the
pipes and persist the model to MLflow, so I could just as well add the
custom non-trainable components there.

**Reply by narayanacharya6** [reply]
Hey @einarbmag , do you mind sharing how you handled persisting to mlflow? Is this just an additional script you run post-training to just push metrics to your tracking server or did you manage to inject something in the training loop itself?
